src/hooks/prisma.ts

```diff
@@ -0,0 +1,139 @@
+import assert from "node:assert";
+
+import { ESTree } from "meriyah";
+import type prisma from "@prisma/client";
+
+import type AppMap from "../AppMap";
+import { getTime } from "../util/getTime";
+import { fixReturnEventIfPromiseResult, recording } from "../recorder";
+import { FunctionInfo } from "../registry";
+import config from "../config";
+
+export default function prismaHook(mod: typeof prisma) {
+  // Imported PrismaClient type does not have _request method in type definition.
+  // But we have it in runtime.
+  assert("PrismaClient" in mod);
+  assert(mod.PrismaClient != null);
+  const PC = mod.PrismaClient as { prototype: unknown };
+  const proto = PC.prototype;
+  assert(proto != null && typeof proto === "object");
+  assert("_request" in proto);
+  proto._request = createProxy(proto._request as (...args: unknown[]) => unknown);
+  return mod;
+}
+
+prismaHook.applicable = function (id: string) {
+  return [config.prismaClientModuleId].includes(id);
+};
+
+// https://github.com/prisma/prisma/blob/095cba1a1b79d0d950246b07c9fb48d22fd7f229/packages/client/src/runtime/getPrismaClient.ts#L181
+interface QueryEvent {
+  timestamp: Date;
+  query: string;
+  params: string;
+  duration: number;
+  target: string;
+}
+
+interface PrismaRequestParamsArgs {
+  data?: unknown;
+  include?: unknown;
+  where?: unknown;
+}
+
+interface PrismaRequestParams {
+  action?: string;
+  model?: string;
+  args?: PrismaRequestParamsArgs;
+}
+
+let hookAttached = false;
+
+const functionInfos = new Map<string, FunctionInfo>();
+
+function getFunctionInfo(model: string, action: string, params: ESTree.Parameter[]) {
+  const key = model + "." + action;
```

Comment:
Is it possible for different calls to the same action to be parametrized differently? If so, this will create just one signature for all of them which will be correct only for the first one. In this case it might make more sense to pass them to the map as a single argument, or explode them all into `(data, include, where)` filling in `undefined` where one isn't provided, to avoid having multiple signatures for the same thing in the AppMap (note they would also need to differ in location because that's how the client code distinguishes code objects).